iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,380 @@
+                // have fewer than max_degrees number of links
+            }
+
+            // now that we know the tree is in a "healthy" state to
+            // recieve more links, add the link to the tree
+            let (cid, bytes) = TreeNode::Leaf(chunk).encode()?;
+            let len = bytes.len();
+            tree[0].push((cid, len));
+            yield (cid, bytes);
+            // at this point, the leaf node may have max_degrees number of
+            // links, but no other stem node will
+        }
+
+        // our stream had 1 chunk that we have already yielded
+        if tree.len() == 1 && tree[0].len() == 1 {
+            return
+        }
+
+        // clean up, aka yield the rest of the stem nodes
+        // since all the stem nodes are able to recieve links
+        // we don't have to worry about "overflow"
+        while let Some(links) = tree.pop_front() {
+            // root node
+            // if tree.len() == 0 {
+            //     let (cid, bytes) = TreeNode::Root(links).encode()?;
+            //     yield (cid, bytes);
+            //     return
+            // }
+
+            let (cid, bytes) = TreeNode::Stem(links).encode()?;
+            let len = bytes.len();
+            yield (cid, bytes);
+
+            if let Some(front) = tree.front_mut() {
+                front.push((cid, len));
+            } else {
+                // final root, nothing to do
+            }
+        }
+    }
+}
+
+fn create_unixfs_node_from_links(links: Vec<(Cid, usize)>, is_root: bool) -> Result<UnixfsNode> {
+    let links = links
+        .into_iter()
+        .map(|(cid, len)| dag_pb::PbLink {
+            hash: Some(cid.to_bytes()),
+            // TODO: notes say that leaf chunks have "", does that mean that
+            // stem nodes should have None?
+            name: Some("".into()),
+            tsize: Some(len as u64),
+        })
+        .collect();
+
+    // PBNode.Data
+    let inner = unixfs_pb::Data {
+        r#type: DataType::File as i32,
+        ..Default::default()
+    };
+
+    // create PBNode
+    let outer = encode_unixfs_pb(&inner, links)?;
+
+    if is_root {
+        return Ok(UnixfsNode::Directory(Node { inner, outer }));
+    }
+
+    // create UnixfsNode
+    Ok(UnixfsNode::File(Node { inner, outer }))
+}
+
+// Leaf and Stem nodes are the two types of nodes that can exist in the tree
+// Leaf nodes encode to `UnixfsNode::Raw`
+// Stem nodes encode to `UnixfsNode::File`
+// Root nodes encode to `UnixfsNode::Directory`
+enum TreeNode {
+    Leaf(Bytes),
+    Stem(Vec<(Cid, usize)>),
+    // Root(Vec<(Cid, usize)>),
+}
+
+impl TreeNode {
+    fn encode(self) -> Result<(Cid, Bytes)> {
+        match self {
+            TreeNode::Leaf(bytes) => {
+                // TODO: notes say that balanced in go-ipfs marks leaves as `File`, not `Raw`, should
+                // I change this?
+                let node = UnixfsNode::Raw(bytes);
```

Comment:
the `merkletree` section of the unixfs notes (https://www.notion.so/number-zero/UnixFs-742339892d9c47d5b79f4f942e661bbf#a2bcc0d96aea4688b27f65eee44a03aa) says that the balanced tree has leafs as `UnixfsNode::File`. Just wanted to double confirm that we are on purpose using `UnixfsNode::Raw`

Reply:
wait just answered my own question. unixfs spec says leaves should be marked as raw (https://github.com/ipfs/specs/blob/main/UNIXFS.md#chunking)

Reply:
both are used in kubo, we are unifying to use Raw encoded leaves